source/agora/consensus/EnrollmentManager.d

```diff
@@ -699,6 +734,100 @@ public class EnrollmentManager
         return true;
     }
 
+    /***************************************************************************
+
+        Check if revealing a pre-image is needed at a certain height
+
+        Params:
+            height = block height to check
+
+        Returns:
+            true if revealing a pre-image is needed
+
+    ***************************************************************************/
+
+    public bool needRevealPreimage (ulong height) @safe nothrow
+    {
+        if (height >= this.next_reveal_height)
+            return true;
+        else
+            return false;
+    }
+
+    /***************************************************************************
+
+        Increase the next reveal height by the revelation period
+
+    ***************************************************************************/
+
+    public void increaseNextRevealHeight () @safe nothrow
+    {
+        ulong next_height = this.getNextRevealHeight();
+        if (this.next_reveal_height < ulong.max)
+            this.setNextRevealHeight(next_height + PreimageRevealPeriod);
+    }
+
+    /***************************************************************************
+
+        Get the next block height to reveal a pre-image
+
+        Returns:
+            the next block height to reveal a pre-image. if any problem in
+            getting the value, it returns the MAX ulong value.
+
+    ***************************************************************************/
+
+    private ulong getNextRevealHeight () @safe nothrow
+    {
+        ulong next_height = ulong.max;
+        try
+        {
+            () @trusted {
+                auto results = this.db.execute("SELECT val FROM node_enroll_data " ~
+                    "WHERE key = ?", "next_reveal_height");
+                if (!results.empty)
+                    next_height = results.oneValue!(size_t);
+            }();
+        }
+        catch (Exception ex)
+        {
+            log.error("Database operation error {}", ex);
```

Comment:
As mentioned before, this is a showstopper error. The node cannot reveal preimages, it should be shut down.

Reply:
> As mentioned before, this is a showstopper error. The node cannot reveal preimages, it should be shut down

There needs to add the code of `assert(0)` in a related PR.